src/crimson/os/seastore/seastore_types.cc

```diff
@@ -239,20 +244,30 @@ std::string device_type_to_string(device_type_t dtype) {
 paddr_t convert_blk_paddr_to_paddr(blk_paddr_t addr, size_t block_size,
     uint32_t blocks_per_segment, device_id_t d_id)
 {
-  paddr_t paddr;
-  paddr.segment = segment_id_t {
+  segment_id_t id = segment_id_t {
     d_id,
-    (uint32_t)(addr / (block_size * blocks_per_segment))
+    (device_segment_id_t)(addr / (block_size * blocks_per_segment))
   };
-  paddr.offset = addr % (block_size * blocks_per_segment);
-  return paddr;
+  segment_off_t off = addr % (block_size * blocks_per_segment);
+  return paddr_t::make_seg_paddr(id, off);
 }
 
 blk_paddr_t convert_paddr_to_blk_paddr(paddr_t addr, size_t block_size,
     uint32_t blocks_per_segment)
 {
-  return (blk_paddr_t)(addr.segment.device_segment_id() *
-	  (block_size * blocks_per_segment) + addr.offset);
+  seg_paddr_t& s = addr.as_seg_paddr();
+  return (blk_paddr_t)(s.get_segment_id().device_segment_id() *
+	  (block_size * blocks_per_segment) + s.get_segment_off());
 }
 
+bool paddr_t::is_null() const {
+  if (get_addr_type() == addr_types_t::SEGMENT) {
+    auto& seg_addr = as_seg_paddr();
+    return seg_addr.get_segment_id() == NULL_SEG_ID ||
+	   dev_addr == make_null().dev_addr;
+  }
```

Comment:
We check for null a lot, so it needs to be inlineable.  Move it into the header as with maybe_relative_to (you do need an inline specifier since it's not inside of the class definition).